hxcite.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXCITE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxcite \- replace bibliographic references by hyperlinks
.SH SYNOPSIS
.B hxcite
.RB "[\| " \-b
.IR base " \|]"
.RB "[\| " \-p
.IR pattern " \|]"
.RB "[\| " \-a
.IR auxfile " \|]"
.RB "[\| " \-m
.IR marker " \|]"
.RB "[\| " -c " \|]"
.IR bibfile " [\| " file " \|]"
.SH DESCRIPTION
.LP
The
.B hxcite
commands copies the
.I file
to standard output, looking for strings of the form [[\fIlabel\fP]].
The label may not include white space and the double pair of square
brackets must enclose the label without any spaces in between. If
.B hxcite
finds the label in the
.IR bibfile ","
the string is replaced by the
.IR pattern "."
The pattern can include certain variables. If the label is not found
in
.IR bibfile ","
it is left unchanged.
.PP
The default pattern replaces the string with a hyperlink, but if the
.B \-p
option is used, the replacement can be any pattern. The input doesn't
even have to be HTML.
.LP
If the label is enclosed in {{...}} instead of [[...]], it is copied
to the output unchanged and not
replaced by the pattern, but the label is still searched in the
.IR bibfile "."
.SH OPTIONS
The following options are supported:
.TP 10
.BI \-p " pattern"
Specifies the pattern by which the string [[\fIlabel\fP]] is replaced.
The pattern may include the variables
.B %b
(which is replaced by the value of the
.B \-b
option),
.B %m
(which is replaced by the value of the
.B \-m
option) and
.B %L
(which is replaced by the
.IR label ")."
The default pattern is
.d
<a href="%b#%L" rel="biblioentry">[%L]<!--{{%m%L}}--></a>
.e
.TP
.BI \-b " base"
Sets the value for the
.B %b
variable in the pattern. Typically this is set to a relative or
absolute URL. By default this value is an empty string.
.TP
.BI \-a " auxfile"
All labels that have been found and replaced are also written to a
file. This is so that
.BR hxmkbib (1)
can find them and create a bibliography. The default
.I auxfile
is constructed from the name of the
.I file
by removing the last extension (if any) and replacing it by ".aux".
If no
.I file
is given, the default name is "aux.aux".
.TP
.BI \-m " marker"
By default, the program looks for "[[name]]", but it can be
made to look for "[[#name]]" where # is some string, usually a
symbol such as '!' or '='. This allows references to be
classified, e.g., "[[!name]]" for normative references and
"[[name]]" for non-normative references.
.TP
.B \-c
Causes "[[name]]" to be ignored when it occurs inside XML comments
("<!--...-->"). This is useful for files where such labels occur in
comments, to avoid that they be expanded and possibly lead to invalid
output; useful also if
.B hxcite
is used for non-HTML files which may contain "<!--" that are not
comment. Occurrences of "{{name}}" are not affected by
.BR \-c "."
(But see warning under BUGS below.)
.SH OPERANDS
The following operands are supported:
.TP 10
.I bibfile
The name of a bibliographic database must be given. It must be a file
in
.BR refer (1)
format and every entry must have at least a
.B %L
field, which is used as label. (Entries without such a field will be
ignored.) Entries may optionally have a
.B %K
line, which should contain whitespace-separated keywords. Those keywords can be used to refer to the entry instead of the label. Thus the
.I foo
in [[\fIfoo\fP]] can either be the label of an entry (%L line) or one
of the keywords of the entry's %K line.
.TP
.I file
The name of the input file is optional. If absent,
.B hxcite
will read from stdin. The file does not have to be an HTML
file, but the default pattern (see the
.B \-p
option) assumes HTML.
.SH "EXIT STATUS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred. Usually this is because a file could not be opened.
Very rarely it may also be an out of memory error.
.SH "VERSIONS"
.PP
March 2000: created by Bert Bos <bert@w3.org> as "cite".
.PP
August 2008: renamed to "hxcite".
.PP
Currently maintained by Bert Bos.
.SH BUGS
.PP
.B hxcite
does not actually parse HTML or XML input and the
.B \-c
option simply treats every occurrence of "<!--" as the start of an XML
comment, even if it occurs inside an attribute value or a CDATA
section. Likewise for "-->" for the end of a comment.
.PP
There is currently no way to use numbers for references (e.g., "[1]",
"[2]") instead of the labels ("[Lie1996]", "[UTN22]").
.PP
.B hxcite
requires the
.B %L
(label) field to be present in every entry in
.IR bibfile ","
which is not the case for
.BR refer "(1)."
.B hxcite
does not implement
.BR refer "'s"
keyword search.
.SH "EXAMPLE"
.PP
The following looks for reference of the form "[[!label]]" in
"myfile.html", skipping references that occur inside HTML comments,
and looks up the labels in "biblio.ref". The output is written to
"new.html" and the list of recognized labels to "myfile.aux".
.d
hxcite -c -m '!' biblio.ref myfile.html > new.html
.e
.SH "SEE ALSO"
.BR asc2xml (1),
.BR refer (1),
.BR hxmkbib (1),
.BR hxnormalize (1),
.BR hxnum (1),
.BR hxprune (1),
.BR hxtoc (1),
.BR hxunent (1),
.BR xml2asc (1),
.BR UTF-8 " (RFC 2279)"


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxcite.c
/*
 * cite - adds hyperlinks to bibliographic references in HTML
 *
 * The programs looks for strings of the form [[name]] (i.e., a
 * bibliographic label inside a double pair of square brackets), e.g.,
 * [[Knuth84]] or [[LieBos97]]. The label will be looked up in a
 * bibliography database and if it is found, the string will be
 * replaced by a pattern which is typically of the form <a
 * href="...">[name]</a>, but the pattern can be changed
 * with a command line option.
 *
 * If the string is of the form {{name}}, the name will be looked up,
 * but the string will be copied unchanged.
 *
 * If the label is not found, a warning is printed and the string is
 * left unchanged.
 *
 * All labels that are found are also stored, one label per line, in a
 * separate file with extension .aux. This file can be used by mkbib
 * to create the bibliography by extracting the corresponding
 * bibliographic entries from the database.
 *
 * The bibliography database must be a refer-style database. Though
 * for the purposes of this program all lines that don't start with
 * "%L" or %K are ignored. Lines with "%L" are assumed to contain a
 * label. Lines with %K are assumed to contain whitespace separated
 * keywords, which are effectively aliases for the label. Entries must
 * have one %L line and one or zero %K lines.
 *
 * Options:
 *
 * -b base
 *     Give the value for %b in the pattern.
 *
 * -p pattern
 *     The replacement for the string [[label]]. The default is
 *
 *     <a href=\"%b#%L\" rel=\"biblioentry\">[%L]<!--{{%m%L}}--></a>
 *
 *     %L will be replaced by the label, %b by the value of the -b
 *     option and %m by the marker (-m option).
 *
 * -a auxfile
 *     The name of the file in which the list of labels will be stored.
 *     Default is the name of the file given as argument, minus its
 *     extension, plus ".aux". If no file is give (input comes from
 *     stdin), the default name is "aux.aux".
 *
 * -m marker
 *     By default, the program looks for "[[name]]", but it can be
 *     made to look for "[[Xname]]" where X is some string, usually a
 *     symbol such as '!' or ='. This allows references to be
 *     classified, e.g., "[[!name]]" for normative references and
 *     "[[name]]" for non-normative references.
 *
 * -c
 *     Assume that every pair "<!--" and "-->" delimit a comment and
 *     do not process any [[label]] that occurs between them. Any
 *     "{{label}}" is processed as normal. This does not actually
 *     parse the input as HTML or XML and thus the program will
 *     mistake occurrences of these two strings inside CDATA sections
 *     or attribute values for comment delimiters.
 *
 * Copyright © 1994-2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 18 March 2000
 * Version: $Id: hxcite.c,v 1.11 2018/02/15 19:02:36 bbos Exp $
 **/

#include "config.h"
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
#include <assert.h>
#include <stdlib.h>
#include <errno.h>
#include <stdio.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
# ifndef HAVE_STRSTR
#  include "strstr.e"
# endif
#endif

#ifdef HAVE_SEARCH_H
#  include <search.h>
#else
#  include "hash.e"
#endif

#include <ctype.h>
#include <stdbool.h>
#include "export.h"
#include "heap.e"
#include "types.e"
#include "errexit.e"


/* Warning: arbitrary limits! */
#define LINESIZE 32768
#define HASHSIZE 4096				/* Size of hash table */

#define WS " \t\r\n\f"				/* Separates %K keywords */

static string base = "";			/* URL of bibilography */
static string mark = "";			/* Flag after "'[[" */
static size_t marklen = 0;			/* Length of mark */
static string prog;				/* = argv[0] */
static string pattern =
  "<a href=\"%b#%L\" rel=\"biblioentry\">[%L]<!--{{%m%L}}--></a>";
static FILE *aux;
static bool skip_comments = false; /* Whether to skip [[ inside <!----> */


/* get_label -- get the label for the keyword, or NULL */
static string get_label(const string keyword)
{
  ENTRY *result, e = {keyword, NULL};

  result = hsearch(e, FIND);
  return result ? (string) result->data : NULL;
}


/* valid_label -- check if the label is well-formed */
static bool valid_label(const string label)
{
  int i;

  for (i = 0; label[i]; i++)
    if (! isalnum(label[i])
	&& label[i] != '-'
	&& label[i] != '_'
	&& label[i] != '.') return false;
  return true;
}


/* expand_ref -- print the reformatted reference */
static void expand_ref(const string label)
{
  int i;

  /* ToDo: somehow allow sequence numbers for references [1], [2], etc. */
  for (i = 0; pattern[i]; i++) {
    if (pattern[i] != '%') {
      putchar(pattern[i]);
    } else {
      switch (pattern[++i]) {
	case '%': putchar('%'); break;		/* Literal '%' */
	case 'b': printf("%s", base); break;	/* Base URL */
	case 'L': printf("%s", label); break;	/* Label */
	case 'm': printf("%s", mark); break;	/* Mark (-m option) */
	default: break;				/* Error in pattern */
      }
    }
  }
}


/* process_line -- look for citations in a line */
EXPORT void process_line(const string text, const string fname, int lineno,
			 bool *in_comment)
{
  string h = text, p, q, label = NULL, key;
  char c;

  /* Loop over occurrences of "[[" + mark + label + "]]"
   and "{{" + mark + label + "}}" */

  while (*in_comment ? (p = strpbrk(h, "-{")) : (p = strpbrk(h, "[{<"))) {

    while (h != p) putchar(*(h++));		/* Print text up to here */

    if (strncmp(p, "-->", 3) == 0) {		/* End of comment */
      putchar(*(h++));
      *in_comment = false;
      continue;
    }
    if (strncmp(p, "<!--", 4) == 0) {		/* Begin of comment */
      putchar(*(h++));
      *in_comment = skip_comments;
      continue;
    }
    if (strncmp(p, "{{", 2) && strncmp(p, "[[", 2)) { /* Not {{ or [[ */
      putchar(*(h++));
      continue;
    }

    /* Is there a corresponding closing bracket? */
    if (! (q = strstr(p + 2, *p == '[' ? "]]" : "}}"))) break;

    c = *p;					/* Remember [ or { */

    if (marklen == 0 || strncmp(p + 2, mark, marklen) == 0) {

      p += 2 + marklen;				/* Skip "[["/"{{" + mark */
      key = newnstring(p, q - p);		/* Extract the key */

      if (! valid_label(key)) {			/* Cannot be a key */
	while (h != q) putchar(*(h++));		/* Copy unchanged */
	putchar(*q); putchar(*(q+1));
      } else if (!(label = get_label(key))) {	/* No citation found: warn */
	while (h != q) putchar(*(h++));		/* Copy unchanged */
	putchar(*q); putchar(*(q+1));
	fprintf(stderr, "%s:%d: warning: no bib entry found for %s\n",
		fname ? fname : (string)"<stdin>", lineno, key);
      } else if (c == '[') {			/* Key found: expand */
	expand_ref(label);			/* Insert full reference */
	fprintf(aux, "%s\n", label);		/* Store label */
      } else {					/* "{{" so don't expand */
	while (h != q) putchar(*(h++));		/* Copy unchanged */
	putchar(*q); putchar(*(q+1));
	fprintf(aux, "%s\n", label);		/* Store label */
      }
      dispose(key);

    } else {					/* No valid mark */

      while (h != q) putchar(*(h++));		/* Copy unchanged */
      putchar(*q); putchar(*(q+1));
    }
    h = q + 2;
  }

  printf("%s", h);				/* Print rest of text */
}


/* store_labels_and_keywords -- store label in hash table */
static void store_labels_and_keywords(const string label, const string keys)
{
  string label1, h, b;
  ENTRY entry;

  assert(label);
  label1 = strtok_r(label, WS, &b);		/* Remove white space */
  if (!label1) return;				/* Empty label */
  entry.key = newstring(label1);
  entry.data = newstring(label1);
  if (!hsearch(entry, ENTER)) errexit("%s: %s\n", prog, strerror(errno));
  if (keys) {
    for (h = strtok_r(keys, WS, &b); h; h = strtok_r(NULL, WS, &b)) {
      entry.key = newstring(h);
      entry.data = newstring(label1);
      if (!hsearch(entry, ENTER)) errexit("%s: %s\n", prog, strerror(errno));
    }
  }
}


/* parse_db -- extract all labels from the refer-style database */
static void parse_db(const string db)
{
  char line[LINESIZE];
  FILE *f;
  int e;
  string label = NULL, keywords = NULL;

  if (!(f = fopen(db,"r"))) errexit("%s: %s: %s\n", prog, db, strerror(errno));

  /* Initialize the hash table */
  if (! hcreate(HASHSIZE)) errexit("%s: %s\n", prog, strerror(errno));

  /* Search for %L lines */
  clearerr(f);
  while (fgets(line, sizeof(line), f)) {
    if (line[0] != '%') {	/* We're outside an entry */
      if (label) store_labels_and_keywords(label, keywords);
      dispose(label);
      dispose(keywords);
    } else if (strncmp(line, "%L ", 3) == 0) {
      label = newstring(line + 3);
    } else if (strncmp(line, "%K ", 3) == 0) {
      keywords = newstring(line + 3);
    }
  }
  if (label) store_labels_and_keywords(label, keywords);

  if ((e = ferror(f))) errexit("%s: %s: %s\n", prog, db, strerror(e));

  if (fclose(f) != 0) errexit("%s: %s: %s\n", prog, db, strerror(errno));
}


/* usage -- print usage message and exit */
static void usage(void)
{
  errexit("Usage: %s [-b base] [-p pattern] [-a auxfile] [-c] [-v] bib-file [HTML-file]\n",
	  prog);
}


int main(int argc, char *argv[])
{
  char line[LINESIZE];
  string h, auxfile = NULL, dbfile = NULL, infile = NULL;
  bool in_comment = false;
  int e, lineno, c;
  FILE *f;

  /* Parse command line arguments */
  prog = argv[0];
  while ((c = getopt(argc, argv, "b:p:a:m:cv")) != -1) {
    switch (c) {
    case 'b': base = optarg; break;		/* Set base of URL */
    case 'p': pattern = optarg; break;		/* Form of expanded ref */
    case 'a': auxfile = optarg; break;		/* Name of auxfile */
    case 'm': mark = optarg; marklen = strlen(mark); break; /* After "[[" */
    case 'c': skip_comments = true; break;	/* Skip [[ in comments */
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    default: usage();
    }
  }
  if (optind == argc || argc > optind + 2) usage();

  dbfile = argv[optind++];
  if (optind != argc) infile = argv[optind++];

  /* Read the labels from the bibliography database */
  parse_db(dbfile);

  /* Construct auxfile */
  if (! auxfile) {
    if (infile) {
      newarray(auxfile, strlen(infile) + 5);
      strcpy(auxfile, infile);
      if ((h = strrchr(auxfile, '.'))) *h = '\0';
      strcat(auxfile, ".aux");
    } else {
      auxfile = "aux.aux";
    }
  }
  if (! (aux = fopen(auxfile, "w")))
    errexit("%s: %s: %s\n", prog, auxfile, strerror(errno));

  /* Open input file or use stdin */
  f = infile ? fopen(infile, "r") : stdin;
  if (!f) errexit("%s: %s: %s\n", prog, infile, strerror(errno));

  /* Read input line by line */
  clearerr(f);
  lineno = 1;
  while (fgets(line, sizeof(line), f))
    process_line(line, infile, lineno++, &in_comment);
  if ((e = ferror(f))) errexit("%s: %s\n", prog, strerror(e));

  fclose(aux);
  fclose(f);
  return 0;
}
